### PR TITLE
port-forward listen on address

### LINKED ISSUE #29678: Add the option to bind to a different address with kubectl port-forward
Even a switch to bind to `0.0.0.0` would do the trick for me.  My use case is debugging when the client needs to be in a Windows VM.

### LINKED ISSUE #36152: kubectl port-forward should allow binding/listening to other interfaces
**What keywords did you search in Kubernetes issues before filing this one?**
port forward bind address
port forward localhost
port forwarding listen

**Is this a BUG REPORT or FEATURE REQUEST? (choose one):** FEATURE REQUEST

**Kubernetes version (use kubectl version):**
```
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.1+33cf7b9", GitCommit:"33cf7b9acbb2cb7c9c72a10d6636321fb180b159", GitTreeState:"not a git tree", BuildDate:"2016-10-13T15:13:44Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:10:32Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Describe *in detail* the feature/behavior/change you'd like to see.**

Currently `kubectl port-forward` only allows binding to localhost...

```
$ kubectl port-forward mypod 4560
Forwarding from 127.0.0.1:4560 -> 4560
Forwarding from [::1]:4560 -> 4560
```

It would be good to be able to bind to another local interface (and "all/*/0.0.0.0"), something like:

```
$ kubectl port-forward --bind-address 0.0.0.0 mypod 4560
Forwarding from 0.0.0.0:4560 -> 4560
Forwarding from [::]:4560 -> 4560
```

```
$ kubectl port-forward --bind-address 192.168.56.1 mypod 4560
Forwarding from 192.168.56.1:4560 -> 4560
```

This would allow more advanced interactions for debugging/interaction between local development environments and containers running on k8s.

[kubectl proxy](http://kubernetes.io/docs/user-guide/kubectl/kubectl_proxy/) supports `--bind-address` which seems like a good a name as any, to be consistent.

See also: https://github.com/kubernetes/kubernetes/issues/25113

### LINKED ISSUE #43962: [proposal] kubectl port-forward listen on custom address
Implementing this proposal that will solve issues raised in https://github.com/kubernetes/kubernetes/issues/36152 and https://github.com/kubernetes/kubernetes/issues/29678. In short: allow binding on addresses other then currently supported `127.0.0.1` and `::1`

The idea is to implement a new method `NewOnAddress` in pkg/client/unversioned/portforward/portforward.go that will add an address parameter, and refactor `New` as a wrapper on the `NewOnAddress` pointing to "localhost" address.

The `NewOnAddress` would work as `New` did previously (that is calling `listenOnPortAndAddress`) with a difference of setting the new PortForwarder private variable `address` that can take string representations of IPv4, IPv6, "localhost" or nil. Then `listenOnPort` will be changed in a way that if IPv4 or IPv6 is provided in `PortForwarder.address`, the address given wil be the only one used (if it is a valid one). If value "localhost" or nil is passed, it will behave exactly as it does now - listening on `127.0.0.1` and `[::1]`.

pkg/kubectl/cmd/portforward.go will be modified to call `NewOnAddress` method instead of `New` with new option address provided by `-a` or `--address` flag on `kubectl port-forward` command and passed via a new string variable `Address` in updated PortForwardOptions type.

### PR DESCRIPTION
**Pull request by goblain**
**What this PR does / why we need it**:

Implements #43962 proposal. Adds `--address` flag to port-forward command that allows listening on addresses other then localhost, so that port-forward can ie. be opened to consumers other then residing in local host like running in docker or different machine/vm

**Which issue this PR fixes**: 

fixes #43962, fixes #36152, fixes #29678

**Release note**:
```
allows selecting non-localhost addresses to listen on with port-forward
```

### CODE REVIEW

**Comment by mengqiy** [comment]
@goblain Are you still working on this? If so, please rebase. I can help review this.

**Comment by goblain** [comment]
@mengqiy please take a look at this

**Comment by mengqiy** [comment]
/ok-to-test
@goblain Thanks for contributing.
Can you add some test please?

**Comment by mengqiy** [comment]
@goblain Any update on the failing test?

**Comment by goblain** [comment]
@mengqiy : sorry for the lag, but I finally refactored this PR to support testing that is in line with existing tests and also added ability to provide multiple comma separated addresses to listen on (including fqdn that will be resolved to the list of IPs to try to listen on)

**Comment by mengqiy** [comment]
Yay, all tests are passing now. I will do a pass soon.

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go
Comment (mengqiy) [inline staging/src/k8s.io/client-go/tools/portforward/portforward.go]: It seems `parsed` will always have one item, why make it a list?

Comment (goblain) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go]: you're completely right, leftover from some refactoring

**Comment by mengqiy** [comment]
/assign @liggitt
PTAL at client-go part.

**Comment by liggitt** [comment]
@smarterclayton what’s the status of switching portforward to websockets from spdy? This seems likely to broaden use of the portforward feature… will the switch involve API level changes we want to make before encouraging broader usage?

**Comment by smarterclayton** [comment]
Yes, we're going to have to introduce a new kubelet to master to client
protocol for web sockets that can handle stream framing in order to handle
multiple connections being forwarded.  I'd like to avoid anything impactful
to port forward until we make that change.

**Comment by jeffmendoza** [comment]
@liggitt @smarterclayton This would be a useful feature to kubectl, and updates?

**Comment by forelabs** [comment]
@goblain please rebase :)

otherwise if nothing will happen, how about do a fork , get those commits and make a new pr?
I rly think this feature is useful.

**Comment by goblain** [comment]
yea, well it kind of got put on hold by expected changes in client, no idea if it's ok to go with it again
@smarterclayton ?

**Comment by chadnickbok** [comment]
Its frustrating that this very reasonable change has been left to die on the vine for months because... No reason really? Some other work was happening and some dude got nervous and decided not to add this right now?

This feature is great and should be evaluated on its merits; as far as I can tell the switch from SPDY to websockets won't impact this code at all.

**Comment by jeffmendoza** [comment]
Bump  @kubernetes/sig-cli-pr-reviews

**Comment by goblain** [comment]
rebased

**Comment by mengqiy** [comment]
> Yes, we're going to have to introduce a new kubelet to master to client
protocol for web sockets that can handle stream framing in order to handle
multiple connections being forwarded.  I'd like to avoid anything impactful
to port forward until we make that change.

@smarterclayton @liggitt What is the status now? Can we merge this? It seems people really like this feature.

**Comment by mengqiy** [comment]
cc: @seans3 @soltysh

**Comment by mrnicoco** [comment]
it's going to be a year now, that's sad. What's missing to be merged ?

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/portforward/portforward.go]: I wouldn't expect to do DNS resolution internally... specifying the actual address (`127.0.0.1`, `::1`, `1.2.3.4`, `0.0.0.0`, etc) is consistent with all other bind address options in kubernetes commands

Comment (goblain) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go]: I must have had a reason to include it, but after all this time I don't really remember. Probably just some minor convenience like not having to script IP4/6 retrieval when you know the name. I am not going to argue much in favor of it though, so if that is a problem, I'll just strip it away.

File: pkg/kubectl/cmd/portforward.go
Comment (liggitt) [inline pkg/kubectl/cmd/portforward.go]: prefer binding to the Address field in the struct, and defaulting to "127.0.0.1", "::1" above when constructing `opts` above

```
cmd.Flags().StringSliceVar(&opts.Addresses, "address", opts.Addresses, "Addresses to listen on (comma separated)")
```

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/portforward/portforward.go]: silently failing on some of the inputs is not what I would expect. I'm unsure whether we want to support the user specifying multiple addresses to bind to... would we expect the command to exit if any were unsuccessful, or if all were unsuccessful? I could see both ways, which makes me think it might be better to limit a single invocation to a single address.

**Comment by liggitt** [comment]
> @smarterclayton:
> Yes, we're going to have to introduce a new kubelet to master to client
protocol for web sockets that can handle stream framing in order to handle
multiple connections being forwarded. I'd like to avoid anything impactful
to port forward until we make that change.

Clayton, what were the current limitations around the port-forward pieces? was it an inability to do multiple concurrent requests? exposing this on non-local addresses seems likely to increase usage impacted by that limitation. Would it make sense to come up with a clear description of the limitation and likely future impact of changes, and mark this flag as experimental so that people who want to use it now are aware of the status and future compatibility guarantees?

File: pkg/kubectl/cmd/util/helpers.go
Comment (liggitt) [inline pkg/kubectl/cmd/util/helpers.go]: these aren't general purpose... remove in favor of direct binding to the opts struct field

Comment (goblain) [reply pkg/kubectl/cmd/util/helpers.go]: sure, in fact it originally was direct, but the changes seemed like now this would be the preferred way. Anyway, will revert to previous way.

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go
Comment (goblain) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go]: at it's current implementation, it binds to two addresses (127.0.0.1, ::1) by default so that behavior should be kept, it is in my opinion perfectly logical to want to listen on particular set of IP addresses and not the others (ie. IP from one external interface and the loopback ones). The command in kubectl explicitly lists ports it successfully bound to

Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go]: > at it's current implementation, it binds to two addresses (127.0.0.1, ::1) by default so that behavior should be kept

sure, if the user doesn't have an address they want to forward on, it should continue binding to "local" addresses.

> it is in my opinion perfectly logical to want to listen on particular set of IP addresses and not the others (ie. IP from one external interface and the loopback ones)

I agree, but if I explicitly want to listen on a set of IP addresses, I would not expect the command to report success if it couldn't accomplish that. Limiting the user-specified bind address to a single address makes it super clear when their intent was not satisfied, and plays well with something using this to script port-forward setups (e.g. loops that restart port-forward on a particular address when it times out or dies)

Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go]: > The argument for scripting "compatibility" makes a lot of sense. I would lean to the idea that it could be implemented with failing if any bind failed

There's the initial bind failure to consider, but also what happens when one of the connections closes, right? I'd expect the following:
* On start, all explicitly requested bind addresses must succeed
* On close of any explicitly requested bind addresses, the entire command should exit

I would stick with the user specifying addresses explicitly (not doing localhost->[127.0.0.1, ::1] expansion for explicit addresses). You can move the defaulting into the command if it makes it easier to distinguish between unspecified addresses (which should preserve best-effort binding to 127.0.0.1 and ::1) and explicitly specified addresses of `127.0.0.1,::1` (which should both succeed or it would be considered failure)

Comment (goblain) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go]: I was looking into it a bit more, and it seems in fact that port-forward is not really built to behave like this at the moment. Look at following vanilla kubectl port-forward : 

    $ kubectl port-forward mysql-server-0 3306:3306 80:3306  
    Forwarding from 127.0.0.1:3306 -> 3306
    Unable to listen on port 80: All listeners failed to create with the following errors: Unable to create listener: Error listen tcp4 127.0.0.1:80: bind: permission denied, Unable to create listener: Error listen tcp6: address [[::1]]:80: missing port in address

So... as long as one of the ports binds correctly, the port-forward stays up and running. 

I'm not saying that is the ultimate argument for not failing if only some of the addresses fail to bind, but to stay consistent, it would stand to reason that we change to fail if any port failed as well, which would be a change from current behavior.

**Comment by goblain** [comment]
@liggitt how about now ?

File: pkg/kubectl/cmd/portforward.go
Comment (lavalamp) [inline pkg/kubectl/cmd/portforward.go]: I don't know if this is still active or not, but it showed up in my email and since I glanced at it, surely this should be called "Interfaces", no?

Comment (goblain) [reply pkg/kubectl/cmd/portforward.go]: /test pull-kubernetes-integration

Comment (goblain) [reply pkg/kubectl/cmd/portforward.go]: As I understand, interfaces would be like `eth0` `wlp3s0` and similar, this is list of IP addresses to listen on

Comment (inc0) [reply pkg/kubectl/cmd/portforward.go]: I agree with @goblain, address != interface and we're dealing with addresses here. Interface can have multiple addresses, we still need to specify which exactly we want to listen on.

**Review by inc0 — APPROVED** [review]

**Comment by davidovich** [comment]
We are also very interested in this feature. What we need to do at the moment is a bit convoluted:

We have a development use-case where we want to share access to our minikube nginx-ingress-controller. Because external connections are not accepted, we are about to script this workaroud:

Start a kubectl proxy on the host running minikube (call it machine A):
```
$ kubectl config set-context minikube
$ kubectl proxy --accept-hosts='.*' --address='0.0.0.0'
```

On the remote machine (machine B), set a kubectl config context to allow access to machine A:

```
$ kubectl config set-cluster machineA --server=http://machineA.corporate.domain:8001 --insecure-skip-tls-verify=true
$ kubectl config set-context machineA-context --cluster=machineA
$ kubectl config use-context machineA-context
$ kubectl --namespace=kube-system port-forward nginx-ingress-controller-2wxt6 8080
```

Then enjoy access to machineA through machineB http://localhost:8080

If `kubectl port-forward` was allowed bind to external traffic, this would be reduced to a simple `kubectl port-forward --address='0.0.0.0'` call. This would also be inline with kubectl proxy.

**Comment by rcoup** [comment]
> What's stopping you currently from invoking kubectl from the host you want the port be available?

For me: credentials and cluster context configuration being copied/passed around unnecessarily. Use case is connecting development VMs, dev runtimes, and dev tools to local (Docker/minikube) + remote k8s clusters

> If you need it, it looks like the built-in unix tools should do that for you, don't they? 

I'm sure socat _can_ do it if you cast the right spells and make the correct sacrifice, but it's not exactly intuitive. Especially to make it listen on everything-other-than-localhost and connect _to_ localhost.

> if this should be implemented I'd imagine one address being specified similarly to how you do that with tools known to me (including docker run), and that would be part of the `port:port`, as in `address:local_port:remote_port`, where address would be optional and defaulted to localhost.

👍

**Comment by dims** [comment]
cc @thockin 

Tim, thoughts please!

**Comment by alluneediswill** [comment]
I would very much look forward to this feature as well. But for anyone who also lands on this page and in need of a solution, here is my workaround with the aforementioned socat
```
kubectl port-forward svc/${SERVICE} 8081:8080 > kubectl_forward.log 2>&1 &
socat TCP4-LISTEN:8080,fork TCP4:localhost:8081
```
It will forward all the traffic to port 8080 from this host to a service in a remote cluster at port 8080.

**Comment by vladimirvivien** [comment]
This is great way to quickly test/validate apps running in pods.  Is there anything that can be done to see this through rather than let it rot?  

Maybe instead of introducing the ability to bind to multiple addresses, maybe keep it simple and just use one address for each invocation of port-forward:

```shell
kubectl port-forward --address 10.19.21.23 pod/mypod 8888:5000
```
Or as suggested above:

```shell
kubectl port-forward  pod/mypod 10.19.21.23:8888:5000
```
Would this (specify 1 address) get it this merged?  

Judging by the number of issues opened, this is something that is useful.  If there is a technical case against doing this at all, it should be stated.

**Comment by raravena80** [comment]
+1

**Comment by Jeeppler** [comment]
+1

**Comment by svenkost** [comment]
+1 I really need this feature.

**Comment by jason008bourne** [comment]
+1 I really need this feature.

**Comment by goblain** [comment]
@mengqiy @liggitt @smarterclayton : what's the status for this? should I rebase in hopes it can be merged anytime soon, or is it still in cold storage for undefined time?

**Comment by smarterclayton** [comment]
We can remove the hold.  The other stuff got delayed, and this won't make it worse.

**Comment by goblain** [comment]
rebased, chances to get lgtm here ?

**Comment by goblain** [comment]
rebased yet again

@dims @mengqiy @liggitt @soltysh chances to get final review and approval / merge before another rebase is required ? :)

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Review by dixudx — COMMENTED** [review]
Actually I was wondering whether we could use `10.19.21.23:8888:5000` to indicate port 8888 open on 10.19.21.23 and forward to 5000 in the pod.

File: pkg/kubectl/cmd/portforward/portforward.go
Comment (lavalamp) [inline pkg/kubectl/cmd/portforward/portforward.go]: I suggest "local-addresses"?

File: pkg/kubectl/cmd/portforward/portforward.go
Comment (lavalamp) [reply pkg/kubectl/cmd/portforward/portforward.go]: LocalAddresses?

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go]: What do you think about changing it to just offer a single address? Users can always run multiple instances of kubectl, and that way they if one fails the user can restart it.

File: pkg/kubectl/cmd/portforward/portforward.go
Comment (lavalamp) [reply pkg/kubectl/cmd/portforward/portforward.go]: (but see my suggestion to only serve one local address)

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go
Comment (lavalamp) [inline staging/src/k8s.io/client-go/tools/portforward/portforward.go]: This looks wrong. I think you need to resolve the name rather than hard coding this stuff.

Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go]: Likely you should attempt resolving any name that isn't an IP address.

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go]: I see there's a good case for running on all loopback addresses, at least.

In that case, I think it's super important that if it can't bind every address requested, it fail-- if something malicious is occupying one of the ports, it'd be a security vuln.

File: pkg/kubectl/cmd/portforward/portforward.go
Comment (goblain) [reply pkg/kubectl/cmd/portforward/portforward.go]: `address` was selected as it's the same as for ie. `kubectl proxy`, so makes sense to stay coherent

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go]: (FTR, I interpret @liggitt's comment as meaning "don't accept localhost as an option, so that resolution isn't necessary".)

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go
Comment (goblain) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go]: as mentioned before, one of the cases where we will fail on one, but not all, is when we make a `localhost` bind on a system that does not support IPv6

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go
Comment (lavalamp) [inline staging/src/k8s.io/client-go/tools/portforward/portforward.go]: Perhaps explicitly say 127.0.0.1 & ::1 here, then you can remove the special case in the parse function?

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go
Comment (lavalamp) [inline staging/src/k8s.io/client-go/tools/portforward/portforward.go]: Document that addresses must all be valid IPv4 or IPv6 addresses, no hostnames?

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go]: Since this PR isn't making things worse, maybe we should merge this and file an issue for followup.

**Comment by lavalamp** [comment]
/approve

I think other changes can be made in a followup.

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go]: In a followup, I think we should do this after confirm that the address exists, if possible.

**Comment by vladimirvivien** [comment]
Wow.  This is awesome.

**Comment by Dimagog** [comment]
This is great!
**Where can I find Windows binary with this change?**
I need it badly and all my attempts to build kubectl myself fail with:
```
E:\Go\home\src\k8s.io\kubernetes\pkg\features\kube_features.go:458:23: cannot use "k8s.io/apiextensions-apiserver/pkg/features".CustomResourceValidation (type "k8s.io/apiextensions-apiserver/vendor/k8s.
io/apiserver/pkg/util/feature".Feature) as type "k8s.io/apiserver/pkg/util/feature".Feature in map key
E:\Go\home\src\k8s.io\kubernetes\pkg\features\kube_features.go:459:23: cannot use "k8s.io/apiextensions-apiserver/pkg/features".CustomResourceSubresources (type "k8s.io/apiextensions-apiserver/vendor/k8
s.io/apiserver/pkg/util/feature".Feature) as type "k8s.io/apiserver/pkg/util/feature".Feature in map key
```

**Comment by mengqiy** [comment]
@Dimagog it should be in this prerelease: https://github.com/kubernetes/kubernetes/releases/tag/v1.13.0-alpha.1

**Comment by Dimagog** [comment]
@mengqiy Yes, latest binary from below has the `--address` switch, thanks!
https://storage.googleapis.com/kubernetes-release/release/v1.13.0-alpha.1/bin/windows/amd64/kubectl.exe

**Pull request merged**
